azurelinuxagent/ga/update.py

```diff
@@ -912,40 +923,101 @@ def _check_and_download_agent_if_upgrade_available(self, protocol, base_version=
         if not conf.get_autoupdate_enabled():
             return False
 
-        now = time.time()
-        if self.last_attempt_time is not None:
-            next_attempt_time = self.last_attempt_time + conf.get_autoupdate_frequency()
-        else:
-            next_attempt_time = now
-        if next_attempt_time > now:
-            return False
+        def report_error(msg_, version=CURRENT_VERSION):
+            logger.warn(msg_)
+            add_event(AGENT_NAME, op=WALAEventOperation.Download, version=version, is_success=False, message=msg_)
 
         family = conf.get_autoupdate_gafamily()
-        logger.info("Checking for agent updates (family: {0})", family)
-
-        self.last_attempt_time = now
-
+        incarnation_changed = False
         try:
-            manifest_list, etag = protocol.get_vmagent_manifests()
-
+            # Fetch the agent manifests from the latest Goal State
+            manifest_list, incarnation = protocol.get_vmagent_manifests()
+            incarnation_changed = self.__goal_state_updated(incarnation)
             manifests = [m for m in manifest_list if m.family == family and len(m.uris) > 0]
             if len(manifests) == 0:
-                logger.verbose(u"Incarnation {0} has no {1} agent updates",
-                               etag, family)
+                logger.verbose(
+                    u"No manifest links found for agent family: {0} for incarnation: {1}, skipping update check".format(
+                        family, incarnation))
                 return False
+        except Exception as err:
+            # If there's some issues in fetching the agent manifests, report it only on incarnation change
+            if incarnation_changed:
```

Comment:
Sorry I dont get your question, the `incarnation_changed` is a boolean, and if the `protocol.get_vmagent_manifests()` throws it'll be caught by this except block. If its the fist GS, this boolean would be true and we would report it. Does that answer your question? If not, can you elaborate it please?

Reply:
you initialized`  incarnation_changed = False` at line #931 and then `protocol.get_vmagent_manifests()` throws before incarnation_changed. So, you would have default `false`, and reporting based on this flag. In this case we don't report right. Correct me if I miss anything here.

Reply:
No that's correct, that was by design. The `protocol.get_vmagent_manifests()` will raise a `ProtocolError` if we were unable to fetch the GS before. In that case, I indeed do not want to report this error as it might end up bombarding our logs.

Reply:
you know what? you got it me here because I just realized that no matter what you do most likely it's going to break with my next change so I guess I'll do this change as well then. good as it is now

just fyi: right now all the test that use the mock protocol hit update_goal_state and use mock data that relies on extensionsConfig. Since we do not have a matching vmSettings for those mock data, we hit a goal state mistmach and we fallback to extensionsConfig, which what those tests need anyway.

this is going to break in my next iteration, when there will be no fallback to extensionsConfig